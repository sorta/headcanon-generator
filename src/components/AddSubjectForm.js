import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddSubjectForm extends Component {
  static propTypes = {
    addSubject: PropTypes.func,
    selectedFandomKey: PropTypes.string,
  };
  static defaultProps = {
    addSubject: () => {},
    selectedFandomKey: '',
  };

  nameRef = React.createRef();

  createSubject = (ev) => {
    ev.preventDefault();
    const subject = {
      name: this.nameRef.current.value,
      fandomKey: this.props.selectedFandomKey,
    };
    this.props.addSubject(this.props.selectedFandomKey, subject);
    ev.currentTarget.reset();
  }

  render() {
    return (
      <form className="AddSubjectForm form-add" onSubmit={this.createSubject}>
        <input name="name" ref={this.nameRef} type="text" placeholder="New Subject Name" required />
        <button type="submit">+</button>
      </form>
    );
  }
}

export default AddSubjectForm;
