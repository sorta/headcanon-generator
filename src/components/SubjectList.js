import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SubjectListItem from './SubjectListItem';

class SubjectList extends Component {
  static propTypes = {
    subjects: PropTypes.object,
    subjectKeys: PropTypes.array,
    updateSubject: PropTypes.func,
    deleteSubject: PropTypes.func,
    isManaging: PropTypes.func,
    setAvailability: PropTypes.func,
    subjectAvailability: PropTypes.object,
    fandomAvailability: PropTypes.object,
  };
  static defaultProps = {
    subjects: {},
    subjectKeys: [],
    updateSubject: () => {},
    deleteSubject: () => {},
    isManaging: () => false,
    setAvailability: () => {},
    subjectAvailability: {},
    fandomAvailability: {},
  };

  render() {
    if (Object.keys(this.props.subjects).length === 0) {
      return null;
    }

    return (
      <ul className="options-list SubjectList">
        {this.props.subjectKeys.sort().map(key => {
          if ({}.hasOwnProperty.call(this.props.subjects, key)) {
            return (
              <SubjectListItem
                key={key}
                index={key}
                subject={this.props.subjects[key]}
                updateSubject={this.props.updateSubject}
                deleteSubject={this.props.deleteSubject}
                isManaging={this.props.isManaging}
                setAvailability={this.props.setAvailability}
                subjectAvailability={this.props.subjectAvailability}
                fandomAvailability={this.props.fandomAvailability}
              />
            );
          } else {
            return null;
          }
        })}
      </ul>
    );
  }
}

export default SubjectList;
