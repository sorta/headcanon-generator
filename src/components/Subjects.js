import React, { Component } from 'react';
import PropTypes from 'prop-types';

import AddSubjectForm from './AddSubjectForm';
import SubjectList from './SubjectList';

class Subjects extends Component {
  static propTypes = {
    fandom: PropTypes.object,
    subjects: PropTypes.object,
    addSubject: PropTypes.func,
    updateSubject: PropTypes.func,
    deleteSubject: PropTypes.func,
    isManaging: PropTypes.func,
    setAvailability: PropTypes.func,
    subjectAvailability: PropTypes.object,
    fandomAvailability: PropTypes.object,
  };
  static defaultProps = {
    fandom: {},
    subjects: {},
    addSubject: () => {},
    updateSubject: () => {},
    deleteSubject: () => {},
    isManaging: () => false,
    setAvailability: () => {},
    subjectAvailability: {},
    fandomAvailability: {},
  };

  render() {
    const currentFandom = this.props.fandom;

    if (typeof currentFandom === 'undefined' || typeof currentFandom.name === 'undefined') {
      return null;
    }

    const subjectKeys = currentFandom.subjects || [];

    return (
      <div className="Subjects">
        <h2 className="Subject-fandom-title">{currentFandom.name}</h2>
        <AddSubjectForm
          addSubject={this.props.addSubject}
          selectedFandomKey={this.props.selectedFandomKey}
          isManaging={this.props.isManaging}
        />
        <SubjectList
          subjects={this.props.subjects}
          subjectKeys={subjectKeys}
          updateSubject={this.props.updateSubject}
          deleteSubject={this.props.deleteSubject}
          isManaging={this.props.isManaging}
          setAvailability={this.props.setAvailability}
          subjectAvailability={this.props.subjectAvailability}
          fandomAvailability={this.props.fandomAvailability}
        />
      </div>
    );
  }
}

export default Subjects;
