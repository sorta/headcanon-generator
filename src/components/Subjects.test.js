import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Subjects from './Subjects';
import AddSubjectForm from './AddSubjectForm';
import SubjectList from './SubjectList';

const props = {
  subjects: {
    s_test_1: { name: 'Test 1' },
    s_test_2: { name: 'Test 2' },
    s_test_3: { name: 'Test 3' },
  },
  subjectAvailability: {},
  updateSubject: jest.fn(),
  deleteSubject: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
  fandomAvailability: {},
  fandom: {},
  addSubject: jest.fn(),
};

// Tests for no fandom selected
describe("Subjects with no fandom selected", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <Subjects { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('is null when no fandom is selected', () => {
    const component = shallow(<Subjects { ...props } />);
    expect(component.getElement()).toBe(null);
  });
});

// Tests for fandom selected
const sProps = { ...props };
sProps.fandom = { name: 'Fandom 1' };
describe("Subjects", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <Subjects { ...sProps } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a title element', () => {
    const component = shallow(<Subjects { ...sProps } />);
    expect(component.find('div.Subjects').exists()).toBe(true);
    expect(component.find('h2.Subject-fandom-title').exists()).toBe(true);
  });

  it('has an AddSubjectForm component', () => {
    const component = shallow(<Subjects { ...sProps } />);
    expect(component.find(AddSubjectForm).exists()).toBe(true);
  });

  it('has a SubjectList component', () => {
    const component = shallow(<Subjects { ...sProps } />);
    expect(component.find(SubjectList).exists()).toBe(true);
  });
});
