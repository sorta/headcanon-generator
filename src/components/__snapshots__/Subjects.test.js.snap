// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subjects renders and matches our snapshot 1`] = `
<div
  className="Subjects"
>
  <h2
    className="Subject-fandom-title"
  >
    Fandom 1
  </h2>
  <ul
    className="options-list SubjectList"
  />
</div>
`;

exports[`Subjects with no fandom selected renders and matches our snapshot 1`] = `null`;
